romio warnings....

Thank you for taking the time to submit an issue!

## Background information

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)

All versuibs

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)



### Please describe the system on which you are running

gcc 7.3.0, gcc 8.2.0

## Details of the problem

These warnings have existed for a long time. I would like them fixed but do not understand what romio is trying to accomplish with this check. This bug is meant to track this issue.

Example warnings:
```
../../../../../../ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c:738:49: warning: comparison of integers of different signs: 'long long' and 'size_t' (aka 'unsigned long') [-Wsign-compare]
      ADIOI_Assert((for_next_iter+coll_bufsize) == (size_t)(for_next_iter+coll_bufsize));
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~  ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../../../ompi/mca/io/romio321/romio/adio/include/adioi.h:1052:24: note: expanded from macro 'ADIOI_Assert'
  #define ADIOI_Assert assert
                       ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
../../../../../../ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c:1032:4: warning: comparison of integers of different signs: 'ADIO_Offset' (aka 'long long') and 'size_t' (aka 'unsigned long') [-Wsign-compare]
                        ADIOI_BUF_COPY
                        ^~~~~~~~~~~~~~
../../../../../../ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c:932:28: note: expanded from macro 'ADIOI_BUF_COPY'
  ADIOI_Assert(size_in_buf == (size_t)size_in_buf); \
               ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
../../../../../../ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c:1039:4: warning: comparison of integers of different signs: 'ADIO_Offset' (aka 'long long') and 'size_t' (aka 'unsigned long') [-Wsign-compare]
                        ADIOI_BUF_COPY
                        ^~~~~~~~~~~~~~
../../../../../../ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c:932:28: note: expanded from macro 'ADIOI_BUF_COPY'
  ADIOI_Assert(size_in_buf == (size_t)size_in_buf); \
               ~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
```

--- a/ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c
+++ b/ompi/mca/io/romio321/romio/adio/common/ad_read_coll.c
@@ -735,7 +735,7 @@ static void ADIOI_Read_and_exch(ADIO_File fd, void *buf, MPI_Datatype
 	if (for_next_iter) {
 	    tmp_buf = (char *) ADIOI_Malloc(for_next_iter);
       ADIOI_Assert((((ADIO_Offset)(MPIU_Upint)read_buf)+real_size-for_next_iter) == (ADIO_Offset)(MPIU_Upint)(read_buf+real_size-for_next_iter));
-      ADIOI_Assert((for_next_iter+coll_bufsize) == (size_t)(for_next_iter+coll_bufsize));
+      ADIOI_Assert((ADIO_Size) (for_next_iter+coll_bufsize) == (size_t)(for_next_iter+coll_bufsize));
 	    memcpy(tmp_buf, read_buf+real_size-for_next_iter, for_next_iter);
 	    ADIOI_Free(fd->io_buf);
 	    fd->io_buf = (char *) ADIOI_Malloc(for_next_iter+coll_bufsize);
@@ -929,7 +929,7 @@ static void ADIOI_R_Exchange_data(ADIO_File fd, void *buf, ADIOI_Flatlist_node
     while (size) { \
 	size_in_buf = ADIOI_MIN(size, flat_buf_sz); \
   ADIOI_Assert((((ADIO_Offset)(MPIU_Upint)buf) + user_buf_idx) == (ADIO_Offset)(MPIU_Upint)((MPIU_Upint)buf + user_buf_idx)); \
-  ADIOI_Assert(size_in_buf == (size_t)size_in_buf); \
+  ADIOI_Assert((ADIO_Size) size_in_buf == (size_t)size_in_buf);          \
 	memcpy(((char *) buf) + user_buf_idx, \
 	       &(recv_buf[p][recv_buf_idx[p]]), size_in_buf); \
 	recv_buf_idx[p] += size_in_buf; /* already tested (size_t)size_in_buf*/ \

--- a/ompi/mca/io/romio321/romio/adio/common/ad_read_str.c
+++ b/ompi/mca/io/romio321/romio/adio/common/ad_read_str.c
@@ -34,7 +34,7 @@
              &status1, error_code); \
         if (*error_code != MPI_SUCCESS) return; \
     } \
-    ADIOI_Assert(req_len == (size_t)req_len); \
+    ADIOI_Assert((ADIO_Size) req_len == (size_t)req_len);                \
     memcpy((char *)buf + userbuf_off, readbuf+req_off-readbuf_off, req_len); \
 }
 

--- a/ompi/mca/io/romio321/romio/adio/common/ad_write_coll.c
+++ b/ompi/mca/io/romio321/romio/adio/common/ad_write_coll.c
@@ -832,7 +832,7 @@ static void ADIOI_W_Exchange_data(ADIO_File fd, void *buf, char *write_buf,
     while (size) { \
         size_in_buf = ADIOI_MIN(size, flat_buf_sz); \
   ADIOI_Assert((((ADIO_Offset)(MPIU_Upint)buf) + user_buf_idx) == (ADIO_Offset)(MPIU_Upint)((MPIU_Upint)buf + user_buf_idx)); \
-  ADIOI_Assert(size_in_buf == (size_t)size_in_buf); \
+  ADIOI_Assert((ADIO_Size) size_in_buf == (size_t)size_in_buf);          \
         memcpy(&(send_buf[p][send_buf_idx[p]]), \
                ((char *) buf) + user_buf_idx, size_in_buf); \
         send_buf_idx[p] += size_in_buf; \

--- a/ompi/mca/io/romio321/romio/adio/common/ad_write_nolock.c
+++ b/ompi/mca/io/romio321/romio/adio/common/ad_write_nolock.c
@@ -335,7 +335,7 @@ void ADIOI_NOLOCK_WriteStrided(ADIO_File fd, const void *buf, int count,
 #ifdef ADIOI_MPE_LOGGING 
 		    MPE_Log_event( ADIOI_MPE_write_a, 0, NULL );
 #endif
-                    ADIOI_Assert(size == (size_t) size);
+                    ADIOI_Assert((ADIO_Size) size == (size_t) size);
                     ADIOI_Assert(off == (off_t) off);
 		    err = write(fd->fd_sys, ((char *) buf) + indx, size);
 #ifdef ADIOI_MPE_LOGGING

--- a/ompi/mca/io/romio321/romio/adio/include/adio.h
+++ b/ompi/mca/io/romio321/romio/adio/include/adio.h
@@ -96,9 +96,11 @@
 
 #ifdef MPI_OFFSET_IS_INT
    typedef int ADIO_Offset;
+   typedef unsigned ADIO_Size;
 #  define ADIO_OFFSET MPI_INT
 #elif defined(HAVE_LONG_LONG_64)
    typedef long long ADIO_Offset;
+   typedef unsigned long long ADIO_Size;
 #  ifdef HAVE_MPI_LONG_LONG_INT
 #     define ADIO_OFFSET MPI_LONG_LONG_INT
 #  else
@@ -106,9 +108,11 @@
 #  endif
 #elif defined(HAVE_INT64)
    typedef __int64 ADIO_Offset;
+   typedef unsigned __int64 ADIO_Size;
 #  define ADIO_OFFSET MPI_DOUBLE
 #else
    typedef long ADIO_Offset;
+   typedef unsigned long ADIO_Size;
 #  define ADIO_OFFSET MPI_LONG
 #endif
 

